src/finch/tensor.py

```diff
@@ -178,60 +209,105 @@ def permute_dims(self, axes: tuple[int, ...]) -> "Tensor":
+        shape = arr_in.shape
+        lvl = Element(arr.dtype.type(fill_value), arr_in.reshape(-1))
+        for size in shape[::-1]:
+            lvl = Dense(lvl, size)
+        return lvl, order
+
+    @classmethod
+    def _from_scipy_sparse(cls, x) -> JuliaObj:
+        if x.format == "coo":
+            return cls.construct_coo_jl_object(
+                coords=(x.col, x.row), data=x.data, shape=x.shape[::-1], order=Tensor.row_major
+            )
+        elif x.format == "csc":
+            return cls.construct_csc_jl_object(
+                arg=(x.data, x.indices, x.indptr),
+                shape=x.shape,
+            )
+        elif x.format == "csr":
+            return cls.construct_csr_jl_object(
+                arg=(x.data, x.indices, x.indptr),
+                shape=x.shape,
+            )
+        else:
+            raise ValueError(f"Unsupported SciPy format: {type(x)}")
 
-class COO(Tensor):
-    def __init__(self, coords, data, shape, fill_value=0.0):
-        assert coords.ndim == 2
+    @classmethod
+    def construct_coo_jl_object(cls, coords, data, shape, order, fill_value=0.0) -> JuliaObj:
+        assert len(coords) == 2
         ndim = len(shape)
+        order = cls.preprocess_order(order, ndim)
 
-        lvl = jl.Element(data.dtype.type(fill_value), jl.Vector(data))
+        lvl = jl.Element(data.dtype.type(fill_value), data)
         ptr = jl.Vector[jl.Int]([1, len(data) + 1])
-        tbl = tuple(jl.Vector(coords[i, :] + 1) for i in range(ndim))
-
-        jl_data = jl.SparseCOO[ndim](lvl, shape, ptr, tbl)
-        jl_data = jl.swizzle(jl.Tensor(jl_data), 1, 2)
+        tbl = tuple(jl.OffByOneVector(arr) for arr in coords)
 
-        super().__init__(jl_data=jl_data)
+        jl_data = jl.swizzle(jl.Tensor(jl.SparseCOO[ndim](lvl, shape, ptr, tbl)), *order)
+        return jl_data
 
+    @classmethod
+    def construct_coo(cls, coords, data, shape, order=row_major, fill_value=0.0) -> "Tensor":
+        return Tensor(jl_data=cls.construct_coo_jl_object(coords, data, shape, order, fill_value))
 
-class _Compressed2D(Tensor):
-    def __init__(self, arg, shape, fill_value=0.0, order=Tensor.row_major):
+    @staticmethod
+    def _construct_compressed2d_jl_object(
+        arg, shape, order, get_permutation_fn, fill_value=0.0,
+    ) -> JuliaObj:
         assert isinstance(arg, tuple) and len(arg) == 3
         assert len(shape) == 2
 
         data, indices, indptr = arg
         dtype = data.dtype.type
-        data = jl.Vector(data)
-        indices = jl.Vector(indices + 1)
-        indptr = jl.Vector(indptr + 1)
+        indices = jl.OffByOneVector(indices)
+        indptr = jl.OffByOneVector(indptr)
 
         lvl = jl.Element(dtype(fill_value), data)
         jl_data = jl.swizzle(
             jl.Tensor(
                 jl.Dense(jl.SparseList(lvl, shape[0], indptr, indices), shape[1])
             ),
-            *self.get_permutation(order),
+            *get_permutation_fn(order),
```

Comment:
I think that CSR should always be row-major and CSC should always be column major. I don't think that we should support an order argument when constructing CSR or CSC. that would eliminate the need for this lambda, and also absolve us of having to talk about "Row major CSC"